distributed/scheduler.py

```diff
@@ -5882,16 +5960,13 @@ def transition(self, key, finish, *args, **kwargs):
                     start,
                     finish2,
                     ts._state,
-                    dict(recommendations),
+                    str(recommendations),
```

Comment:
Any reason for this str conversion?

Reply:
There seems to be a fair bit of preventive copying in this code. I'm assuming this is out of concern that `recommendations` might get unintentionally modified. Thought that as this is building a logging message and `str` is immutable this would maintain the intent of the original code while avoiding copying the `dict`. Am ok just dropping the copy and using `recommendations` if that is preferred

Reply:
I was mostly curious what triggers this change. Which approach is best depends a bit on how these logs are used. For instance, the `{Scheduler,Worker}.story` uses these to filter the messages for keys. I believe the filtering there uses some kind of set logic, so I would assume we'll need to adjust it there as well.
I find working with these things programatically a lot easier in a structured container like a dict. If we want it to be immutable, something like `tuple(recommendations.items())` would preserve the structure, or at least would allow easier programatic access w/out needing to parse `str(dict)`. reverting this is surprisingly difficult :)

I would also expect the str conversion to be more expensive than the dict copy. A shallow copy should only copy the hash table and since the recommendations are mostly small-ish dicts I would hope the shallow copy is ok. otoh, a str conversion would need to copy the entire dict content, wouldn't it?

Reply:
Yeah I don't have strong feelings here if you would prefer it reverted. Was just thinking copying to a `dict` might not be needed. Am ok dropping the `str(...)` bit or reverting, whichever you prefer.

Reply:
Went ahead and just reverted this